### ISSUE TITLE
investigate Q3's AAS system return

### ISSUE THREAD

**Issue by bmorel** [Unvanquished/Unvanquished#1866]
Through my journey in unv's code, and more specifically through the AI's code, I started to think the paradigm we use are currently ok, even if badly implemented, except for one: `RecastNavigation`.

In a nutshell, RecastNavigation is a wanna-be-generic navigation system for bots. It includes several tools:

1. creation of navmeshes for capsule-shaped agents: Q3 and hence unvanquished uses axial-BBoxes
2. navigation through said 2D navmesh: Q3 uses lot of jumping and special moves, and same for unvanquished
3. crowd control if all actors are the same size: unvanquished actors are almost never the same ones, either capabilities or shape or both differs.

One thing have to be said, and explains the Post-Beta milestone: the navmesh generator *currently works*. It have many problems, but it do works.
RecastNavigation is the only FOSS library to implement this AFAIK, and it's documentation is rather limited.
Sure, it does have a decent API reference. I mean it. One still needs to read the code, so it's only decent, and not good. And the API still warns a lot. But that's ok.
Still, it does not have actual documentation. Only demos. And API ref.

On the contrary, Q3's AAS have no lib AFAIK (not searched yet) but it has an old thesis (from 2001) explaining how it works, so, no reference API but actual documentation.

There might be several still alive implementations around, and unvanquished's engine is meant to still use Q3's BSP map system.

AAS felt optimised, in my readings, for the BSP map format. Is that surprising, for a 90's software? Not much.
But it also took into considerations things like rocket jumps, mentioning those were costly. We certainly don't have rocket jumps in quake's meaning, but the pounces are quite alike. About the cost, they were costly... on a 650MHz P3.

The bigger con for that change is: it's not implemented yet. Hence the post-bêta milestone.

But the pros, I think, are heavy weight:

* my lesser-auto-jump hack (IIRC I have a security to prevent jumps over 45% of agent's height...)? No more needed, instead we'd benefit from actual jump-pathfining
* about water? No more problem. That works for jetpack, too.
* I even think it's doable to implement wallwalk navigation. What is certain though is that it's possible to *automatically handle ladders* something I can't even think about a hack to get automated with recast!

**Comment by bmorel** [comment]
Note that those problems might be fixed:

[ ] #1673 (no navmesh in AAS)
[ ] #1448 (Q3's AAS uses BSP, so as long as map is not buggy...)
[ ] #1220 (Q3's bots are even able to use moving platforms, said @illwieckz )
[ ] #1672 (can be done now, but would also be done with less calculations with AAS since all the precomp tables... ok, not a strong pro)
[ ] #1574 because AAS would make it easier to build "heatmaps" of danger (@necessarily-equal )
[ ]  #1493 because it simply would remove the requirement for so many traces
[ ] #310 because it would allow a bot to know when they can jump down without danger
[ ] #1862 because. Just because. Recast could do it, too, but in both cases a deep study is requires, and I don't think recast's study is worth it considering the other points.
[ ] #1863 because AAS is even more dynamic than recast (which have a tiiiiny bit of dynamism)
[ ] #1865 would be usable to distract bots from the A* algo. Actually, AAS would allow us finer control on pathfinding, to just avoid many problems but I'm moving out of topic here.
[ ] #1867 would no longer be hacks. They're normal features of AAS. And this blows my mind. Really. I could even add jumpads and teleporters support that navmeshes are unable to support in any FOSS implementation I'm aware of.

**Comment by slipher** [comment]
Amusingly, I found a thread with a guy talking about how AAS sucks and he is moving to the greener grass of Recast: https://www.doomworld.com/forum/topic/112159-quake-3-bots-using-navmesh/. The thread seems to imply that there is an open-source AAS implementation.

**Comment by bmorel** [comment]
Well, the post in question does not explains the why of the changes.
My personal only references for the idea that AAS might be better than recast are based on the q3 thesis (that this thread mentions) and the limited understanding I have about recast.
For a game like counter-strike, recast is good enough (as long as maps do not have water, that is, but I don't remember such maps when I was at school, not that I ever was a big fan of CS...) because players only have normal movement abilities: no teleporters, no jumpers, no wallwalking, no wallrunning... so the navmesh technology can work.

In a game with complex movements or map mechanics like what we have in Unvanquished, or other games like red-eclipse (1&2), xonotic... the movements are a lot in 3D, and a 2.5D navigation system can not easily allow to leverage automatically all the mechanics.
And this is even worst for red-eclipse 1 or unvanquished, as the movements those games allow are *really* complex: both games allow to wall-jump, and unvanquished even allows wallwalking.

For things like teleportation support, [the limitation is only in recast implementation](https://groups.google.com/g/recastnavigation/c/8lV6uzSFnz8/m/hf4oWdTVQJ8J), not in the navmesh technology. Also, [some people managed to get impressive results](https://groups.google.com/g/recastnavigation/c/oDH5659Bz5w/m/G3vb_a1HXRIJ) in other areas, and there's code provided there.
None of those things lead to any way to get bots do underwater navigation or to allow for climbing, as shown by title of the last link: "Agents cannot "jump" down / Portals / Pathfinding". You will note that we can already (manually) build jump-down information... but unvanquished uses a lot of "jump-up" moves, and on that area, the only improvement I could get was the "lesser auto-jump" feature, which is, when generating navmeshes, use STEPSIZE+part of the jump magnitude (IIRC I capped it at ~60% of agent's size, because of bugs, but maybe I was wrong and the bug I had was unrelated to this... I'll need to re-evaluate). Despite this helps a lot (example: nexus6 default human base often got stuck bots, this is fixed. In Perseus, bots would often take a long path in H base instead of jumping up/down, also fixed. In old parpax, there was small "traps": similarly fixed.) this is not perfect, a notable example being transit station, in the ruined area: bots there are stuck until they go dragoon or battlesuit, which allows them to unstuck.

So, while I know several people seem to dislike AAS, I have yet to see any real complaint about it or limits of a real implementation, while I am aware of the problems (for our use-case!) recast have, and some of those are in the core of the concept.
Simply put, recastnavigation would be perfect to build a wheeled robot, but not for a flying drone, and unvanquished *do* needs the flying drone abilities.
